crates/ruff_python_semantic/src/binding.rs

```diff
@@ -549,71 +549,64 @@ bitflags! {
 /// A trait for imported symbols.
 pub trait Imported<'a> {
     /// Returns the call path to the imported symbol.
-    fn call_path(&self) -> &[&'a str];
+    fn qualified_name(&self) -> &QualifiedName<'a>;
 
     /// Returns the module name of the imported symbol.
     fn module_name(&self) -> &[&'a str];
 
     /// Returns the member name of the imported symbol. For a straight import, this is equivalent
     /// to the qualified name; for a `from` import, this is the name of the imported symbol.
     fn member_name(&self) -> Cow<'a, str>;
-
-    /// Returns the fully-qualified name of the imported symbol.
-    fn qualified_name(&self) -> String {
-        let mut output = String::new();
-        format_qualified_name_segments(self.call_path(), &mut output).unwrap();
-        output
-    }
 }
 
 impl<'a> Imported<'a> for Import<'a> {
     /// For example, given `import foo`, returns `["foo"]`.
-    fn call_path(&self) -> &[&'a str] {
-        self.qualified_name.as_ref()
+    fn qualified_name(&self) -> &QualifiedName<'a> {
+        &self.qualified_name
     }
 
     /// For example, given `import foo`, returns `["foo"]`.
     fn module_name(&self) -> &[&'a str] {
-        &self.qualified_name[..1]
+        &self.qualified_name.segments()[..1]
     }
 
     /// For example, given `import foo`, returns `"foo"`.
     fn member_name(&self) -> Cow<'a, str> {
-        Cow::Owned(self.qualified_name())
+        Cow::Owned(self.qualified_name().to_string())
     }
 }
 
 impl<'a> Imported<'a> for SubmoduleImport<'a> {
     /// For example, given `import foo.bar`, returns `["foo", "bar"]`.
-    fn call_path(&self) -> &[&'a str] {
-        self.qualified_name.as_ref()
+    fn qualified_name(&self) -> &QualifiedName<'a> {
+        &self.qualified_name
     }
 
     /// For example, given `import foo.bar`, returns `["foo"]`.
     fn module_name(&self) -> &[&'a str] {
```

Comment:
It would be nice if we could return `QualifiedNameRef` or similar from here because the module name is a qualified name too. But I consider this out of the scope of this PR.